rfcs/2020-06-18-2625-architecture-revisit.md

```diff
@@ -0,0 +1,504 @@
+transform into source tasks does not work out, the same interface changes would
+allow alternative options (e.g. a threadpool) with no change to components
+themselves.
+
+The concept of pipelines per source task gives us a concurrency model that
+scales naturally with load while taking advantage of natural partitioning of the
+data. It can be extended later with some user-facing idea of partitioning, but
+we can start to reap the benefits of partitioning without putting that burden on
+users.
+
+Sources with embedded pipelines will very quickly start to look like a single
+topology node with multiple outputs. By modelling this specifically in our
+internals, we have a natural place to start experimenting with programmable
+control flow. This experience and groundwork will be invaluable if/when we
+choose to add it as a user-facing feature.
+
+Formalizing the idea of a staged, compiler-like path from raw config files to
+running topologies gives us the opportunity to clean up and better organize
+a number of existing features. Environment variable interpolation, multiple
+config files, type checking, etc are are examples of potential stages that could
+be much better implemented in this framework. Other requested features like
+snippets, input wildcards, etc would also be much easier to implement with this
+in place.
+
+Introducing different types of sinks (e.g. streaming vs request-response) gives
+us the opportunity to consolidate shared behavior from sinks of the same type.
+Right now, service-based sinks all do their own work to build up middleware
+stacks for rate limiting, retries, concurrency, etc. This could be pushed up
+into topology building if we have enough information to differentiate these
+sinks for those that do not need these wrapping layers. It also opens up the
+possibility of using these sinks in alternative runtime setups, like a hosted
+HTTP service. Instead of being wrapped in a `Sink` -like structure, services
+could be called directly from an HTTP handler.
+
+## Drawbacks
+
+There are a number of downsides to the changes proposed here:
+
+1. We are adding both implementation complexity as well as increasing the size
+   of the mental model needed to understand Vector's behavior.
+
+2. It's possible that tying computational concurrency to source task concurrency
+   results will work well in some situations and work poorly in others. It adds
+   a significant level of variability to something that used to be relatively
+   consistent. This applies to both performance as well as memory use.
+
+3. Many of the benefits here will not come automatically. We'll need to make
+   sure that we keep the overall design in mind and make individual components
+   that work well with it. This applies particularly to sources that don't
+   currently run more than one task (e.g. file source, UDP-based sources, etc).
+
+4. It's not entirely clear how runtimes like Lua and WASM should fit into this
```

Comment:
WASM modules have a two way handshake protocol where we can both establish unidirectional settings and validate bidrectional settings. I think we can adapt to be able to use the new protocols based on what the wasm module registers itself as accepting.